### ISSUE TITLE
Compatibility with VisiumHD 

### ISSUE THREAD

**Issue by Rafael-Silva-Oliveira** [BayraktarLab/cell2location#358]
Following up with this: https://github.com/BayraktarLab/cell2location/issues/356

Are there any improvements coming soon in cell2location to take in the full data for very large datasets? The approach of splitting in batches removes any spatial information that might be used to do cell deconvolution so taking in the full dataset would be important

With the new VisiumHD being more prominent, the number of spots or bins can range anywhere from 150k to 650k+, so when I test with cell2location, I always get memory errors.

And doing in batches also doesn't work very well (the full dataset would have to be passed along). 

![image](https://github.com/BayraktarLab/cell2location/assets/70116258/77eb46c8-2c3e-41c1-9da5-474b72d3145e)

**Comment by vitkl** [comment]
Solving modelling challenges for VisiumHD, NovaST, OpenST and other similar high-resolution technologies is a research project rather than a matter of tweaking models to work on larger data. Subcellular resolution of measurements requires rethinking what it means to analyse this data. I see two major ways forward:
1. Modelling observations at high resolution but ignoring cell boundaries. We could estimate cell type and subcellular gene programme densities at 1-2um resolution using CNN, GNN, conditional autoregressive models as approaches to pool information at the subcellular resolution. You also need to represent that scRNA-derived RNA abundance profile needs to be decomposed into several subcellular compartments. I implemented a CNN-based approach that considers these factors in this branch https://github.com/BayraktarLab/cell2location/pull/337 but it likely requires much more compute than people in this community are used to (many GPUs many days). This approach is also related to https://github.com/HiDiHlabs/ssam by @pjb7687.
2. Using cell segmentation by modifying the Baysor approach to work with 1000s of genes. Maybe @VPetukhov can weigh in too.

If anyone is interested in collaborating to make this happen please reach out to me, Omer and Oliver by email.

**Comment by vitkl** [comment]
A practical way to analyse VisiumHD right now is to aggregate it at 8um or 20um or 50um resolution (depending on data quality) and using cell2location with tips discussed in #356.

**Comment by Rafael-Silva-Oliveira** [comment]
> A practical way to analyse VisiumHD right now is to aggregate it at 8um or 20um or 50um resolution (depending on data quality) and use cell2location with tips discussed in #356.

Hey thank you so much for your valuable input! Indeed, this was something I was thinking - Performing clustering in order to achive 25-55 micron artificial "spots" and perform deconvolution on these. Another thing I was testing was cell typist that is commonly used for scRNA annotation, and since the slots are so small now, it's no longer a manner of spot deconvolution anymore. So by using the 8 micron resolution and assuming that each bin is a cell, cell typist seemed to do okay in that mapping, but I'm still figuring out the biological meaning of that mapping (if there's one to begin with). Other would be to use cell typist to annotate the 2 micron, then use these predictions to predict the 8 micron bin with majority voting, and then use these 8 micron bins to predict predict the 16 micron bins with majority voting once again. But then again, the goal of this new HD technology would be to avoid performing deconvolution as that was one of the main problems to solve from that technology, so I'm also interested in finding a solution to this for my research

**Comment by Rafael-Silva-Oliveira** [comment]
> > A practical way to analyse VisiumHD right now is to aggregate it at 8um or 20um or 50um resolution (depending on data quality) and use cell2location with tips discussed in #356.
> 
> Hey thank you so much for your valuable input! Indeed, this was something I was thinking - Performing clustering in order to achive 25-55 micron artificial "spots" and perform deconvolution on these. Another thing I was testing was cell typist that is commonly used for scRNA annotation, and since the slots are so small now, it's no longer a manner of spot deconvolution anymore. So by using the 8 micron resolution and assuming that each bin is a cell, cell typist seemed to do okay in that mapping, but I'm still figuring out the biological meaning of that mapping (if there's one to begin with). Other would be to use cell typist to annotate the 2 micron, then use these predictions to predict the 8 micron bin with majority voting, and then use these 8 micron bins to predict predict the 16 micron bins with majority voting once again. But then again, the goal of this new HD technology would be to avoid performing deconvolution as that was one of the main problems to solve from that technology, so I'm also interested in finding a solution to this for my research

Another thing from these more probabilistic methods is that it's heavily dependent on the intersection between the genes present in the scRNA used as reference and the ST data. So even when I use the 16 bin resolution, which has over 150k bins for a sample I'm using, I'm essentially only using 1500-2000 genes to do the "predictions" on these 150k bins

**Comment by LinearParadox** [comment]
Another possible improvement might come on the torch's end. Not sure how this would impact scvi tools, but cuda currently has a way to treat system memory as gpu memory. Pytorch doesn't have this implemented to my knowledge (https://github.com/pytorch/pytorch/issues/104417#issue-1781120079) and https://developer.nvidia.com/blog/simplifying-gpu-application-development-with-heterogeneous-memory-management/.

**Comment by vitkl** [comment]
Visium HD data can be segmented into cells using https://github.com/Teichlab/bin2cell and then counts aggregated into segmented cells can be analysed using cell2location to examine cell purity and to further decompose areas with spatially interlaced cell types.

**Comment by Rafael-Silva-Oliveira** [comment]
> Visium HD data can be segmented into cells using https://github.com/Teichlab/bin2cell and then counts aggregated into segmented cells can be analysed using cell2location to examine cell purity and to further decompose areas with spatially interlaced cell types.

I've been using bin2cell, but even then it produces quite a few number of cells (one of the datasets I've been testing still counts 350-400k cells out of the 9 million + bins of 2 micron resolution) so cell2location would still throw some memory errors I'm afraid

Right now Bin2Cell + CellTypist seems to be the most scalable approach for cell type inference using a scRNA reference

**Comment by lisch7** [comment]
@vitkl @Rafael-Silva-Oliveira Hello, I am interested in using bin2cell to segment cells and then applying cell2loc for further analysis. I followed the demo notebook (https://nbviewer.org/github/Teichlab/bin2cell/blob/main/notebooks/demo.ipynb) of bin2cell but I don't know how to do next when I finished the pipline. Could you please guide me if there are any tutorials or reproducible code available that demonstrate this workflow? I would prefer not to use CellTypist as I already have a fully annotated single-cell dataset. Thank you!

**Comment by cafferychen777** [comment]
Hi everyone,

Thank you for this valuable discussion. Cell2location has been an incredibly influential tool in the spatial transcriptomics field, and we've learned a lot from its elegant probabilistic framework.

For those still looking for solutions to handle very large datasets (100k+ spots), we recently developed **FlashDeconv** as a complementary approach specifically optimized for computational efficiency. It uses randomized linear algebra to achieve linear O(N) scaling, which allows processing ~1M spots in about 3 minutes on a standard workstation.

We would really appreciate it if anyone facing these scalability challenges could give it a try and share feedback - we're still actively improving it and user input would be invaluable:

- GitHub: https://github.com/cafferychen777/flashdeconv
- Paper: https://doi.org/10.64898/2025.12.22.696108
- Install: `pip install flashdeconv`

Of course, for smaller datasets or when the full Bayesian inference is preferred, cell2location remains an excellent choice with its rigorous uncertainty quantification.

Thanks again to @vitkl and the team for all their work on this problem!
